### ISSUE TITLE
Start of computation misinterpreted when mixing multiple: variable and statement function declarations.

### ISSUE THREAD

**Issue by piotrows** [ecmwf-ifs/loki#104]
In some cases, two or more header files with statement function declarations need to be included in the kernel. With preprocessing enabled, the include files are injected into kernel code. At the moment Loki may misinterpret such code containing pattern:
1) variable declaration
2) statement function declaration
3) variable declaration
4) statement function declaration
so the vector_section is inserted before 4), resulting in the Fortran code with statement function declarations inside the loop.
```# (C) Copyright 2018- ECMWF.
# This software is licensed under the terms of the Apache Licence Version 2.0
# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
# In applying this licence, ECMWF does not waive the privileges and immunities
# granted to it by virtue of its status as an intergovernmental organisation
# nor does it submit to any jurisdiction.

import pytest

from loki import (
    Subroutine, Dimension, fgen, Sourcefile, SubroutineItem
)
from transformations import (
    SCCDevectorTransformation, SCCHoistTransformation
)
from conftest import available_frontends
#pylint: disable=too-many-lines

@pytest.fixture(scope='module', name='horizontal')
def fixture_horizontal():
    return Dimension(name='horizontal', size='nlon', index='jl', bounds=('start', 'end'))


@pytest.fixture(scope='module', name='vertical')
def fixture_vertical():
    return Dimension(name='vertical', size='nz', index='jk')


@pytest.fixture(scope='module', name='blocking')
def fixture_blocking():
    return Dimension(name='blocking', size='nb', index='b')



@pytest.mark.parametrize('frontend', available_frontends())
def test_scc_hoist_multiple_kernels(frontend, horizontal, vertical, blocking):
    """
    Test hoisting of column temporaries to "driver" level.
    """

    fcode_driver = """
  SUBROUTINE column_driver(nlon, nz, q, nb)
    INTEGER, INTENT(IN)   :: nlon, nz, nb  ! Size of the horizontal and vertical
    REAL, INTENT(INOUT)   :: q(nlon,nz,nb)
    REAL :: T(NLON)
    INTEGER :: b, start, end

    start = 1
    end = nlon
    do b=1, nb
      call compute_column2(start, end, nlon, nz, q(:,:,b))
    end do
  END SUBROUTINE column_driver
"""

    fcode_kernel2 = """
  SUBROUTINE compute_column2(start, end, nlon, nz, q)
    INTEGER, INTENT(IN) :: start, end  ! Iteration indices
    INTEGER, INTENT(IN) :: nlon, nz    ! Size of the horizontal and vertical
    REAL, INTENT(INOUT) :: q(nlon,nz)
    REAL :: t(nlon,nz)
    INTEGER :: jl, jk
    REAL :: c

REAL(KIND=JPRB) :: FOEDELTA
REAL(KIND=JPRB) :: PTARE,RTT
FOEDELTA (PTARE) = MAX (0.0_JPRB,SIGN(1.0_JPRB,PTARE-RTT))

REAL(KIND=JPRB) :: FESW
REAL(KIND=JPRB) :: PTARET,R2ES,R3LES,RTT,R4LES

FESW(PTARET) = R2ES * EXP(R3LES*(PTARET-RTT)/(PTARET-R4LES))


    c = 5.345
    DO jk = 2, nz
      DO jl = start, end
        t(jl, jk) = c * k
        q(jl, jk) = q(jl, jk-1) + t(jl, jk) * c
      END DO
    END DO

    ! The scaling is purposefully upper-cased
    DO JL = START, END
      Q(JL, NZ) = Q(JL, NZ) * C
    END DO
  END SUBROUTINE compute_column2
"""
    driver_item = SubroutineItem(name='#column_driver',
                                 source=Sourcefile.from_source(fcode_driver, frontend=frontend))
    kernel2 = Subroutine.from_source(fcode_kernel2, frontend=frontend)
    driver = Subroutine.from_source(fcode_driver, frontend=frontend)
    driver.enrich_calls(kernel2)  # Attach kernel source to driver call
    scc_transform = (SCCDevectorTransformation(horizontal=horizontal),)
    scc_transform += (SCCHoistTransformation(horizontal=horizontal, vertical=vertical,
                                             block_dim=blocking),)
    for transform in scc_transform:
        transform.apply(driver, role='driver',
                        targets=['compute_column2'],
                        item=driver_item)
        transform.apply(kernel2, role='kernel')
    print(fgen(kernel2))
    assert False
```
As a result of this test, the following code is printed:
```SUBROUTINE compute_column2 (start, end, nlon, nz, q, t, jl)
  INTEGER, INTENT(IN) :: start, end  ! Iteration indices
  INTEGER, INTENT(IN) :: nlon, nz  ! Size of the horizontal and vertical
  REAL, INTENT(INOUT) :: q(nlon, nz)
  REAL, INTENT(INOUT) :: t(nlon, nz)
  INTEGER :: jk
  REAL :: c

  REAL(KIND=JPRB) :: FOEDELTA
  REAL(KIND=JPRB) :: PTARE, RTT
  FOEDELTA(PTARE) = MAX(0.0_JPRB, SIGN(1.0_JPRB, PTARE - RTT))

  REAL(KIND=JPRB) :: FESW
  REAL(KIND=JPRB) :: PTARET, R2ES, R3LES, RTT, R4LES
  INTEGER, INTENT(IN) :: jl
vector_section FESW(PTARET) = R2ES*EXP(R3LES*(PTARET - RTT) / (PTARET - R4LES))


  c = 5.345
  DO jk=2,nz
    t(jl, jk) = c*k
    q(jl, jk) = q(jl, jk - 1) + t(jl, jk)*c
  END DO

  ! The scaling is purposefully upper-cased
  Q(JL, NZ) = Q(JL, NZ)*C
END SUBROUTINE compute_column2```

**Comment by mlange05** [comment]
Ok, I can confirm that this is actually a problem in the frontends, not the SCC transofrmation; the issues is that the second statement function is not detected as such. Instead of adding it add the end of the `routine.spec` the FP frontend will add this as a plain `Assignment` node to the beginning of the `routine.body`. I'll try and add a test to the frontend and fix this.

**Comment by mlange05** [comment]
Proposed fix submitted as PR #132
